README.md
# cordova-plugin-app-scanner
Using a camera to scan QR codes or barcodes
## Installation

	cordova plugin add cordova-plugin-app-scanner

## UnInstall
	cordova plugin rm cordova-plugin-app-scanner

### Supported Platforms

- Android
- iOS




### Quick Example
	scan.recognize(function (success) {
      	console.log(success);
    }, function (error) {
     	console.log(error);
    });


RELEASENOTES
<!--
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
# 
# http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
#  KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#
-->
# Release Notes
### 1.0.1 (May 9, 2023)
* removed extraneous statement
* Update static resource files


src/ios/CDVScan.m
//
//  CDVScan.m
//  cordova-plugin-scan
//
//  Created by ZhangJian on 16/8/3.
//  Copyright © 2016年 zhangjian. All rights reserved.
//

#import "CDVScan.h"

#import <Cordova/CDVViewController.h>
#import "CDVScan.h"
#import "CDVScanViewController.h"
@interface CDVScan () {
    CDVInvokedUrlCommand *_command;
}
@end

@implementation CDVScan
-(void)pluginInitialize{
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(handleScan:) name:@"scan" object:nil];
}

-(void)handleScan:(NSNotification*)ns{
    CDVPluginResult* pluginResult = nil;
    pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:ns.userInfo[@"content"]];
    [self.commandDelegate sendPluginResult:pluginResult callbackId:_command.callbackId];
}


- (void)recognize:(CDVInvokedUrlCommand*)command{
    // 将传入的参数保存到成员变量中
    _command = command;
    // 创建一个导航控制器，并将扫描控制器作为根视图控制器
    UINavigationController *navController =[[UINavigationController alloc] initWithRootViewController:[[CDVScanViewController alloc] init]];
    // 设置导航栏的背景颜色为透明色
    [navController.navigationBar setBackgroundImage:[UIImage new] forBarMetrics:UIBarMetricsDefault];
    [navController.navigationBar setShadowImage:[UIImage new]];
    // 设置导航栏标题的文本属性
    [navController.navigationBar setTitleTextAttributes:@{NSForegroundColorAttributeName:[UIColor whiteColor]}];
    // 设置导航栏按钮的颜色
    [navController.navigationBar setTintColor:[UIColor whiteColor]];
    // 设置导航栏是否半透明
    [navController.navigationBar setTranslucent:YES];
    // 设置导航控制器全屏显示
    navController.modalPresentationStyle = UIModalPresentationFullScreen;
    // 在当前窗口的根视图控制器上展示导航控制器
    [[[[[UIApplication sharedApplication] delegate] window] rootViewController] presentViewController:navController animated:YES completion:NULL];
}

- (void) dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}
@end



src/android/src/com/thomas/BroadcastDemo/Scan.java
package com.thomas.BroadcastDemo;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.json.JSONArray;
import org.json.JSONException;
import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;



public class Scan extends CordovaPlugin {
    private final static int SCANNIN_GREQUEST_CODE = 1;
    private CallbackContext callbackContext;

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {

        Activity activity = cordova.getActivity();
        this.callbackContext = callbackContext;


        if("recognize".equals(action)) {

            Intent intent = new Intent();
            intent.setClass(activity, MipcaActivityCapture.class);
            cordova.startActivityForResult((CordovaPlugin) this,intent, SCANNIN_GREQUEST_CODE);
        }
        return true;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        switch (requestCode) {
            case SCANNIN_GREQUEST_CODE:
                if(resultCode == Activity.RESULT_OK){
                    Bundle bundle = data.getExtras();
                    callbackContext.success(bundle.getString("result"));
                }
                break;
        }
    }

}


src/android/src/com/thomas/view/ViewfinderView.java
/*
 * Copyright (C) 2008 ZXing authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.thomas.view;

import java.util.Collection;
import java.util.HashSet;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.Typeface;
import android.graphics.drawable.BitmapDrawable;
import android.util.AttributeSet;
import android.view.View;

import com.google.zxing.ResultPoint;
import com.thomas.zxing.camera.CameraManager;

/**
 * This view is overlaid on top of the camera preview. It adds the viewfinder
 * rectangle and partial transparency outside it, as well as the laser scanner
 * animation and result points.
 *
 */
public final class ViewfinderView extends View {
	private static final String TAG = "log";

	private static final long ANIMATION_DELAY = 10L;
	private static final int OPAQUE = 0xFF;


	private int ScreenRate;


	private static final int CORNER_WIDTH = 4;
	// private static final int CORNER_WIDTH = 10;

	private static final int MIDDLE_LINE_WIDTH = 6;


	private static final int MIDDLE_LINE_PADDING = 5;


	private static final int SPEEN_DISTANCE = 5;


	private static float density;

	private static final int TEXT_SIZE = 16;

	private static final int TEXT_PADDING_TOP = 30;


	private Paint paint;


	private int slideTop;


	private int slideBottom;


	private Bitmap resultBitmap;
	private final int maskColor;
	private final int resultColor;

	private final int resultPointColor;
	private Collection<ResultPoint> possibleResultPoints;
	private Collection<ResultPoint> lastPossibleResultPoints;

	boolean isFirst;
	private Context context;

	public ViewfinderView(Context context, AttributeSet attrs) {
		super(context, attrs);
		this.context = context;
		density = context.getResources().getDisplayMetrics().density;

		ScreenRate = (int)(50 * density);
		// ScreenRate = (int)(20 * density);

		paint = new Paint();
		Resources resources = getResources();
		// maskColor = resources.getColor(R.color.viewfinder_mask);
		maskColor = Color.parseColor("#60000000");

		// resultColor = resources.getColor(R.color.result_view);
		resultColor = Color.parseColor("#b0000000");

		resultPointColor = Color.parseColor("#c0ffff00");
		// resultPointColor = resources.getColor(R.color.possible_result_points);
		possibleResultPoints = new HashSet<ResultPoint>(5);
	}

	@Override
	public void onDraw(Canvas canvas) {

		Rect frame = CameraManager.get().getFramingRect();
		if (frame == null) {
			return;
		}

		if(!isFirst){
			isFirst = true;
			slideTop = frame.top;
			slideBottom = frame.bottom;
		}


		int width = canvas.getWidth();
		int height = canvas.getHeight();

		paint.setColor(resultBitmap != null ? resultColor : maskColor);

		canvas.drawRect(0, 0, width, frame.top, paint);
		canvas.drawRect(0, frame.top, frame.left, frame.bottom + 1, paint);
		canvas.drawRect(frame.right + 1, frame.top, width, frame.bottom + 1,
				paint);
		canvas.drawRect(0, frame.bottom + 1, width, height, paint);



		if (resultBitmap != null) {
			// Draw the opaque result bitmap over the scanning rectangle
			paint.setAlpha(OPAQUE);
			canvas.drawBitmap(resultBitmap, frame.left, frame.top, paint);
		} else {


			paint.setColor(Color.GREEN);
			/*
			canvas.drawRect(frame.left, frame.top, frame.left + ScreenRate,
					frame.top + CORNER_WIDTH, paint);
			canvas.drawRect(frame.left, frame.top, frame.left + CORNER_WIDTH, frame.top
					+ ScreenRate, paint);
			canvas.drawRect(frame.right - ScreenRate, frame.top, frame.right,
					frame.top + CORNER_WIDTH, paint);
			canvas.drawRect(frame.right - CORNER_WIDTH, frame.top, frame.right, frame.top
					+ ScreenRate, paint);
			canvas.drawRect(frame.left, frame.bottom - CORNER_WIDTH, frame.left
					+ ScreenRate, frame.bottom, paint);
			canvas.drawRect(frame.left, frame.bottom - ScreenRate,
					frame.left + CORNER_WIDTH, frame.bottom, paint);
			canvas.drawRect(frame.right - ScreenRate, frame.bottom - CORNER_WIDTH,
					frame.right, frame.bottom, paint);
			canvas.drawRect(frame.right - CORNER_WIDTH, frame.bottom - ScreenRate,
					frame.right, frame.bottom, paint);
			*/

			int scanWidth = 50;
			Rect lineRectLT = new Rect();
			lineRectLT.left = frame.left;
			lineRectLT.right = frame.left + scanWidth;
			lineRectLT.top = frame.top;
			lineRectLT.bottom = frame.top + scanWidth;
			// canvas.drawBitmap(((BitmapDrawable)(getResources().getDrawable(R.drawable.scan_left_top))).getBitmap(), null, lineRectLT, paint);
			canvas.drawBitmap(((BitmapDrawable)(getResources().getDrawable(getId("scan_left_top","drawable")))).getBitmap(), null, lineRectLT, paint);


			Rect lineRectRT = new Rect();
			lineRectRT.left = frame.right - scanWidth;
			lineRectRT.right = frame.right;
			lineRectRT.top = frame.top;
			lineRectRT.bottom = frame.top + scanWidth;
			// canvas.drawBitmap(((BitmapDrawable)(getResources().getDrawable(R.drawable.scan_right_top))).getBitmap(), null, lineRectRT, paint);
			canvas.drawBitmap(((BitmapDrawable)(getResources().getDrawable(getId("scan_right_top","drawable")))).getBitmap(), null, lineRectRT, paint);


			Rect lineRectLB = new Rect();
			lineRectLB.left = frame.left;
			lineRectLB.right = frame.left + scanWidth;
			lineRectLB.top = frame.bottom -scanWidth;
			lineRectLB.bottom = frame.bottom ;
			// canvas.drawBitmap(((BitmapDrawable)(getResources().getDrawable(R.drawable.scan_left_bottom))).getBitmap(), null, lineRectLB, paint);
			canvas.drawBitmap(((BitmapDrawable)(getResources().getDrawable(getId("scan_left_bottom","drawable")))).getBitmap(), null, lineRectLB, paint);


			Rect lineRectRB = new Rect();
			lineRectRB.left = frame.right - scanWidth;
			lineRectRB.right = frame.right;
			lineRectRB.top = frame.bottom -scanWidth;
			lineRectRB.bottom = frame.bottom;
			canvas.drawBitmap(((BitmapDrawable)(getResources().getDrawable(getId("scan_right_bottom","drawable")))).getBitmap(), null, lineRectRB, paint);
			// canvas.drawBitmap(((BitmapDrawable)(getResources().getDrawable(R.drawable.scan_right_bottom))).getBitmap(), null, lineRectRB, paint);




			slideTop += SPEEN_DISTANCE;
			if(slideTop >= frame.bottom){
				slideTop = frame.top;
			}
			// canvas.drawRect(frame.left + MIDDLE_LINE_PADDING, slideTop - MIDDLE_LINE_WIDTH/2, frame.right - MIDDLE_LINE_PADDING,slideTop + MIDDLE_LINE_WIDTH/2, paint);
			Rect lineRect = new Rect();
			lineRect.left = frame.left;
			lineRect.right = frame.right;
			lineRect.top = slideTop;
			// lineRect.bottom = slideTop + 18;
			lineRect.bottom = slideTop + 3;
			// canvas.drawBitmap(((BitmapDrawable)(getResources().getDrawable(R.drawable.scan_center_line))).getBitmap(), null, lineRect, paint);
			canvas.drawBitmap(((BitmapDrawable)(getResources().getDrawable(getId("scan_center_line","drawable")))).getBitmap(), null, lineRect, paint);

			paint.setColor(Color.WHITE);
			paint.setTextSize(TEXT_SIZE * density);
			paint.setAlpha(0x40);
			paint.setTypeface(Typeface.create("System", Typeface.BOLD));
			// String textPileCode = getResources().getString(R.string.scan_pile_code);
			// String textPileCode = getResources().getString(getId("scan_pile_code","string"));
			String textPileCode = getResources().getString(getId("scan_text","string"));
			float textWidth = paint.measureText(textPileCode);
			canvas.drawText(textPileCode, frame.left+(frame.right-frame.left-textWidth)/2, (float) ( 50 + frame.bottom + (float)TEXT_PADDING_TOP *density), paint);



			/*Collection<ResultPoint> currentPossible = possibleResultPoints;
			Collection<ResultPoint> currentLast = lastPossibleResultPoints;
			if (currentPossible.isEmpty()) {
				lastPossibleResultPoints = null;
			} else {
				possibleResultPoints = new HashSet<ResultPoint>(5);
				lastPossibleResultPoints = currentPossible;
				paint.setAlpha(OPAQUE);
				paint.setColor(resultPointColor);
				for (ResultPoint point : currentPossible) {
					canvas.drawCircle(frame.left + point.getX(), frame.top
							+ point.getY(), 6.0f, paint);
				}
			}
			if (currentLast != null) {
				paint.setAlpha(OPAQUE / 2);
				paint.setColor(resultPointColor);
				for (ResultPoint point : currentLast) {
					canvas.drawCircle(frame.left + point.getX(), frame.top
							+ point.getY(), 3.0f, paint);
				}
			}*/



			postInvalidateDelayed(ANIMATION_DELAY, frame.left, frame.top,
					frame.right, frame.bottom);

		}
	}

	private int getId(String idName,String type){
		return context.getResources().getIdentifier(idName, type, context.getPackageName());
	}
	public void drawViewfinder() {
		resultBitmap = null;
		invalidate();
	}

	/**
	 * Draw a bitmap with the result points highlighted instead of the live
	 * scanning display.
	 *
	 * @param barcode
	 *            An image of the decoded barcode.
	 */
	public void drawResultBitmap(Bitmap barcode) {
		resultBitmap = barcode;
		invalidate();
	}

	public void addPossibleResultPoint(ResultPoint point) {
		possibleResultPoints.add(point);
	}

}


src/ios/CDVScanViewController.m
//
//  CDVScanViewController.m
//  cordova-plugin-scan
//
//  Created by ZhangJian on 16/8/3.
//  Copyright © 2016年 zhangjian. All rights reserved.
//

#import "CDVScanViewController.h"
#import <AVFoundation/AVFoundation.h>
#import <QuartzCore/QuartzCore.h>

#define QRCodeWidth  260.0   //正方形二维码的边长
#define SCREENHeight  [UIScreen mainScreen].bounds.size.height
#define SCREENWidth  [UIScreen mainScreen].bounds.size.width
static const CGFloat kBorderW = 100;
static const CGFloat kMargin = 30;
@interface CDVScanViewController () <AVCaptureMetadataOutputObjectsDelegate> //用于处理采集信息的代理
// MARK: - Properties
@property (strong, nonatomic) AVCaptureSession* scanSession; //输入输出的中间桥梁
@property (strong, nonatomic) AVCaptureVideoPreviewLayer* scanLayer; //输入输出的中间桥梁
@end
@implementation CDVScanViewController

// MARK: - Lifecycle
- (void)viewDidLoad
{
    [super viewDidLoad];
    self.title = @"扫描二维码/条形码";
    UIButton *leftButton = [[UIButton alloc] initWithFrame:CGRectMake(0, 0, 25, 25)];
    [leftButton setImage:[UIImage imageNamed:@"scanSources.bundle/backArrow"] forState:UIControlStateNormal];
    [leftButton addTarget:self action:@selector(handleClose:) forControlEvents:UIControlEventTouchUpInside];
    UIBarButtonItem *leftItem = [[UIBarButtonItem alloc] initWithCustomView:leftButton];
    self.navigationItem.leftBarButtonItems = [NSArray arrayWithObjects:leftItem, nil];
    //    self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"关闭" style:UIBarButtonItemStylePlain target:self action:@selector(handleClose:)];
    
    [self setupMaskView];
    
    [self setupScanWindowView];
    
    [self beginScanning];
}

- (void)viewWillAppear:(BOOL)animated
{
    if (!self.scanSession.isRunning) {
        [self.scanSession startRunning];
    }
}

- (void)viewDidDisappear:(BOOL)animated
{
    if (self.scanSession.isRunning) {
        [self.scanSession stopRunning];
    }
}

// MARK: - Private Methods
- (void)setupMaskView{
    
    //操作提示
    UILabel * tipLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, SCREENHeight*0.8-kBorderW*2, SCREENWidth, kBorderW)];
    tipLabel.text = @"请将二维码/条形码放入框内, 即可自动识别";
    tipLabel.textColor = [UIColor whiteColor];
    tipLabel.textAlignment = NSTextAlignmentCenter;
    tipLabel.lineBreakMode = NSLineBreakByWordWrapping;
    tipLabel.numberOfLines = 0;
    tipLabel.font = [UIFont systemFontOfSize:15];
    tipLabel.backgroundColor = [UIColor clearColor];
    [self.view addSubview:tipLabel];
    
    //设置统一的视图颜色和视图的透明度
    UIColor *color = [UIColor blackColor];
    float alpha = 0.3;
    
    //设置扫描区域外部上部的视图
    UIView *topView = [[UIView alloc]init];
    topView.frame = CGRectMake(0, 0, SCREENWidth, (SCREENHeight-QRCodeWidth)/2.0-64);
    topView.backgroundColor = color;
    topView.alpha = alpha;
    
    //设置扫描区域外部左边的视图
    UIView *leftView = [[UIView alloc]init];
    leftView.frame = CGRectMake(0, topView.frame.size.height, (SCREENWidth-QRCodeWidth)/2.0,QRCodeWidth);
    leftView.backgroundColor = color;
    leftView.alpha = alpha;
    
    //设置扫描区域外部右边的视图
    UIView *rightView = [[UIView alloc]init];
    rightView.frame = CGRectMake((SCREENWidth-QRCodeWidth)/2.0+QRCodeWidth,topView.frame.size.height, (SCREENWidth-QRCodeWidth)/2.0,QRCodeWidth);
    rightView.backgroundColor = color;
    rightView.alpha = alpha;
    
    //设置扫描区域外部底部的视图
    UIView *botView = [[UIView alloc]init];
    botView.frame = CGRectMake(0, QRCodeWidth+topView.frame.size.height,SCREENWidth,SCREENHeight-QRCodeWidth-topView.frame.size.height);
    botView.backgroundColor = color;
    botView.alpha = alpha;
    
    //将设置好的扫描二维码区域之外的视图添加到视图图层上
    [self.view addSubview:topView];
    [self.view addSubview:leftView];
    [self.view addSubview:rightView];
    [self.view addSubview:botView];
    
    //闪光灯
    UIButton * flashBtn=[UIButton buttonWithType:UIButtonTypeCustom];
    flashBtn.frame = CGRectMake((SCREENWidth-40)/2,SCREENHeight*0.9-kBorderW*2+80, 40, 40);
    [flashBtn setBackgroundImage:[UIImage imageNamed:@"scanSources.bundle/sgd"] forState:UIControlStateNormal];
    flashBtn.contentMode=UIViewContentModeScaleAspectFit;
    [flashBtn addTarget:self action:@selector(openFlash:) forControlEvents:UIControlEventTouchUpInside];
	flashBtn.alpha = alpha;
    [self.view addSubview:flashBtn];
}

- (void)setupScanWindowView{
    //设置扫描区域的位置
    UIView *scanWindow = [[UIView alloc] initWithFrame:CGRectMake((SCREENWidth-QRCodeWidth)/2.0,(SCREENHeight-QRCodeWidth)/2.0-64,QRCodeWidth,QRCodeWidth)];
    scanWindow.clipsToBounds = YES;
    [self.view addSubview:scanWindow];
    
    //设置扫描区域的动画效果
    CGFloat scanNetImageViewH = 2;
    CGFloat scanNetImageViewW = scanWindow.frame.size.width;
    UIImageView *scanNetImageView = [[UIImageView alloc]initWithImage:[UIImage imageNamed:@"scanSources.bundle/scanLine"]];
    scanNetImageView.frame = CGRectMake(0, -scanNetImageViewH, scanNetImageViewW, scanNetImageViewH);
    CABasicAnimation *scanNetAnimation = [CABasicAnimation animation];
    scanNetAnimation.keyPath =@"transform.translation.y";
    scanNetAnimation.byValue = @(QRCodeWidth);
    scanNetAnimation.duration = 2.0;
    scanNetAnimation.repeatCount = MAXFLOAT;
    [scanNetImageView.layer addAnimation:scanNetAnimation forKey:nil];
    [scanWindow addSubview:scanNetImageView];
    
    //设置扫描区域的四个角的边框
    CGFloat buttonWH = 18;
    UIButton *topLeft = [[UIButton alloc]initWithFrame:CGRectMake(0,0, buttonWH, buttonWH)];
    [topLeft setImage:[UIImage imageNamed:@"scanSources.bundle/left-up"]forState:UIControlStateNormal];
    [scanWindow addSubview:topLeft];
    
    UIButton *topRight = [[UIButton alloc]initWithFrame:CGRectMake(QRCodeWidth - buttonWH,0, buttonWH, buttonWH)];
    [topRight setImage:[UIImage imageNamed:@"scanSources.bundle/right-up"]forState:UIControlStateNormal];
    [scanWindow addSubview:topRight];
    
    UIButton *bottomLeft = [[UIButton alloc]initWithFrame:CGRectMake(0,QRCodeWidth - buttonWH, buttonWH, buttonWH)];
    [bottomLeft setImage:[UIImage imageNamed:@"scanSources.bundle/left-down"]forState:UIControlStateNormal];
    [scanWindow addSubview:bottomLeft];
    
    UIButton *bottomRight = [[UIButton alloc]initWithFrame:CGRectMake(QRCodeWidth-buttonWH,QRCodeWidth-buttonWH, buttonWH, buttonWH)];
    [bottomRight setImage:[UIImage imageNamed:@"scanSources.bundle/right-down"]forState:UIControlStateNormal];
    [scanWindow addSubview:bottomRight];
}

- (void)beginScanning{
    // Do any additional setup after loading the view, typically from a nib.
    //获取摄像设备
    AVCaptureDevice* device = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
    //创建输入流
    AVCaptureDeviceInput* input = [AVCaptureDeviceInput deviceInputWithDevice:device error:nil];
    if (!input) return;
    //创建输出流
    AVCaptureMetadataOutput* output = [[AVCaptureMetadataOutput alloc] init];
    //设置代理 在主线程里刷新
    [output setMetadataObjectsDelegate:self queue:dispatch_get_main_queue()];
    
    //初始化链接对象
    self.scanSession = [[AVCaptureSession alloc] init];
    //高质量采集率
    [self.scanSession setSessionPreset:AVCaptureSessionPresetHigh];
    
    [self.scanSession addInput:input];
    [self.scanSession addOutput:output];
    //设置扫码支持的编码格式(如下设置条形码和二维码兼容)
    output.metadataObjectTypes = @[ AVMetadataObjectTypeQRCode, AVMetadataObjectTypeEAN13Code, AVMetadataObjectTypeEAN8Code, AVMetadataObjectTypeCode128Code, AVMetadataObjectTypeCode93Code ];
    //设置扫描有效范围
    //特别注意的地方：有效的扫描区域，定位是以设置的右顶点为原点。屏幕宽所在的那条线为y轴，屏幕高所在的线为x轴
    CGFloat x = ((SCREENHeight-QRCodeWidth)/2.0)/SCREENHeight;
    CGFloat y = ((SCREENWidth-QRCodeWidth)/2.0)/SCREENWidth;
    CGFloat width = QRCodeWidth/SCREENHeight;
    CGFloat height = QRCodeWidth/SCREENWidth;
    output.rectOfInterest = CGRectMake(x, y, width, height);
    //    [output setRectOfInterest:CGRectMake(124/SCREENHeight, ((SCREENWidth-220)/2)/SCREENWidth, 220/SCREENHeight, 220/SCREENWidth)];
    
    self.scanLayer = [AVCaptureVideoPreviewLayer layerWithSession:self.scanSession];
    self.scanLayer.videoGravity = AVLayerVideoGravityResizeAspectFill;
    [self.view.layer insertSublayer:self.scanLayer atIndex:0];
    //开始捕获
    [self.scanSession startRunning];
}

-(void)handleClose:(id)sender{
    [self.navigationController dismissViewControllerAnimated:YES completion:NULL];
}

- (void)viewDidLayoutSubviews
{
    [super viewDidLayoutSubviews];
    self.scanLayer.frame = self.view.layer.bounds;
}


- (void)captureOutput:(AVCaptureOutput*)captureOutput didOutputMetadataObjects:(NSArray*)metadataObjects fromConnection:(AVCaptureConnection*)connection
{
    [self.scanSession stopRunning];
    if (metadataObjects.count > 0) {
        //[session stopRunning];
        AVMetadataMachineReadableCodeObject* metadataObject = [metadataObjects objectAtIndex:0];
        //输出扫描字符串
        NSString *str = metadataObject.stringValue;
        NSLog(@"%@",str);
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            [[NSNotificationCenter defaultCenter] postNotificationName:@"scan" object:self userInfo:@{@"content":str}];
            [self handleClose:NULL];
        });
        
        
    }
}

//MARK:-  闪光灯
-(void)openFlash:(UIButton*)button{
    NSLog(@"闪光灯");
    button.selected = !button.selected;
    if (button.selected) {
        [self turnTorchOn:YES];
    }
    else{
        [self turnTorchOn:NO];
    }
}

//MARK:- 开关闪光灯
- (void)turnTorchOn:(BOOL)on{
    Class captureDeviceClass = NSClassFromString(@"AVCaptureDevice");
    if (captureDeviceClass != nil) {
        AVCaptureDevice *device = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
        if ([device hasTorch] && [device hasFlash]){
            [device lockForConfiguration:nil];
            if (on) {
                [device setTorchMode:AVCaptureTorchModeOn];
                [device setFlashMode:AVCaptureFlashModeOn];
                
            } else {
                [device setTorchMode:AVCaptureTorchModeOff];
                [device setFlashMode:AVCaptureFlashModeOff];
            }
            [device unlockForConfiguration];
        }
    }
}

@end
